File: frontend/src/services/api/statistics.ts
import { apiClient } from './client';
import type { ApiResponse } from '../../types';

export interface PieDatum {
  id: string;
  label: string;
  value: number;
}

export interface LinePoint {
  x: string;
  y: number;
}

export interface LineSeries {
  id: string;
  data: LinePoint[];
}

export interface StatisticsSummary {
  totalSpent: number;
  totalOrders: number;
  averageOrderValue: number;
  mostPurchasedCategory?: string | null;
  mostUsedPlatform?: string | null;
}

export interface CategoryStatistics {
  share: PieDatum[];
  monthlyTrend: LineSeries[];
}

export interface PlatformStatistics {
  ratio: PieDatum[];
  monthlyTrend: LineSeries[];
}

export interface PatternStatistics {
  hourlyTrend: LineSeries[];
  monthlyTotal: LineSeries[];
}

export interface StatisticsDashboardData {
  userId: number;
  summary: StatisticsSummary;
  category: CategoryStatistics;
  platform: PlatformStatistics;
  pattern: PatternStatistics;
}

const mapSummary = (raw: any): StatisticsSummary => ({
  totalSpent: raw?.total_spent ?? 0,
  totalOrders: raw?.total_orders ?? 0,
  averageOrderValue: raw?.average_order_value ?? 0,
  mostPurchasedCategory: raw?.most_purchased_category ?? null,
  mostUsedPlatform: raw?.most_used_platform ?? null,
});

export const statisticsApi = {
  getDashboard: async (user_id: number): Promise<ApiResponse<{ statistics: StatisticsDashboardData }>> => {
    const response = await apiClient.get<any>('/statistics/dashboard', { user_id });

    if (!response.success || !response.data) {
      return {
        success: false,
        error: response.error ?? '통계 데이터를 불러오지 못했습니다.',
        data: null as any
      };
    }

    const data: StatisticsDashboardData = {
      userId: response.data.user_id,
      summary: mapSummary(response.data.summary),
      category: {
        share: response.data.category?.share ?? [],
        monthlyTrend: response.data.category?.monthly_trend ?? [],
      },
      platform: {
        ratio: response.data.platform?.ratio ?? [],
        monthlyTrend: response.data.platform?.monthly_trend ?? [],
      },
      pattern: {
        hourlyTrend: response.data.pattern?.hourly_trend ?? [],
        monthlyTotal: response.data.pattern?.monthly_total ?? [],
      },
    };

    return {
      success: true,
      data: { statistics: data },
    };
  },
};

File: backend/app/models/statistics.py
"""Pydantic schemas for statistics API responses."""

from __future__ import annotations

from pydantic import BaseModel


class PieDatum(BaseModel):
    id: str
    label: str
    value: float


class LinePoint(BaseModel):
    x: str
    y: float


class LineSeries(BaseModel):
    id: str
    data: list[LinePoint]


class StatisticsSummary(BaseModel):
    total_spent: float
    total_orders: int
    average_order_value: float
    most_purchased_category: str | None = None
    most_used_platform: str | None = None


class CategoryStatistics(BaseModel):
    share: list[PieDatum]
    monthly_trend: list[LineSeries]


class PlatformStatistics(BaseModel):
    ratio: list[PieDatum]
    monthly_trend: list[LineSeries]


class PatternStatistics(BaseModel):
    hourly_trend: list[LineSeries]
    monthly_total: list[LineSeries]


class StatisticsDashboardResponse(BaseModel):
    user_id: int
    summary: StatisticsSummary
    category: CategoryStatistics
    platform: PlatformStatistics
    pattern: PatternStatistics
